collector/admission/boundedqueue.go

```diff
@@ -0,0 +1,148 @@
+package admission
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	orderedmap "github.com/wk8/go-ordered-map/v2"
+)
+
+type BoundedQueue struct {
+	maxLimitBytes int64
+	maxLimitWaiters int64
+	currentBytes int64
+	currentWaiters int64
+	lock sync.Mutex
+	// waiters waiters 
+	waiters *orderedmap.OrderedMap[uuid.UUID, waiter]
+}
+
+type waiter struct {
+	readyCh chan struct{} 
+	pendingBytes int64
+	ID uuid.UUID 
+}
+
+func NewBoundedQueue(maxLimitBytes, maxLimitWaiters int64) *BoundedQueue {
+	return &BoundedQueue{
+		maxLimitBytes: maxLimitBytes,
+		maxLimitWaiters: maxLimitWaiters,
+		waiters: orderedmap.New[uuid.UUID, waiter](), 
+	}
+}
+
+func (bq *BoundedQueue) admit(pendingBytes int64) (bool, error) {
+	bq.lock.Lock()
+	defer bq.lock.Unlock()
+
+	if pendingBytes > bq.maxLimitBytes { // will never succeed
+		return false, fmt.Errorf("rejecting request, request size larger than configured limit")
+	}
+
+	if bq.currentBytes + pendingBytes <= bq.maxLimitBytes { // no need to wait to admit
+		bq.currentBytes += pendingBytes
+		return true, nil
+	}
+
+	// since we were unable to admit, check if we can wait.
+	if bq.currentWaiters + 1 > bq.maxLimitWaiters { // too many waiters
+		return false, fmt.Errorf("rejecting request, too many waiters")
```

Comment:
Suggest defining a top-level (and using it in the test). I suggest that for this error (and not the other two) because this is common and something a caller might test against, whereas the other two are irregularities/bugs.

```
const ErrTooManyWaiters = fmt.Errorf("rejecting request, too many waiters")
```

Reply:
Makes sense thanks!